### ISSUE TITLE
[Feature Request] Support more popular compression algorithms and highly optimized kernels on the CPU

### ISSUE THREAD

**Issue by yiliu30** [stochasticai/xTuring#264]
xTuring, is known for its efficient and straightforward fine-tuning support for popular LLMs, but it missing features related to popular compression algorithms, particularly weight-only quantization. These compression methods have been widely acknowledged for their efficiency and are commonly adopted in the industry. Furthermore, xTuring has limited support for CPU-side optimization.
Our team developed these quantization algorithms on [Intel® Neural Compressor](https://github.com/intel/neural-compressor#intel-neural-compressor) and [Intel-Extension-for-Transformers](https://github.com/intel/intel-extension-for-transformers). We want to integrate these algorithms into the xTuring. This integration aims to:

- Offer weight-only quantization algorithms, including [RTN](https://arxiv.org/abs/2306.00978), [AWQ](https://arxiv.org/abs/2306.00978), and TEQ with ignorable accuracy loss 
- Provide optimized kernels for accelerating inference, especially on Intel CPUs

### Usage

```python
from xturing.models import BaseModel

# Sepcific the quantizatuion configuration
from intel_extension_for_transformers.transformers import WeightOnlyQuantConfig
woq_config = WeightOnlyQuantConfig(weight_dtype='int8')
model = BaseModel.create("gpt2", quantization_config=woq_config)

# Inference model with itrex's highly optimized kernels
output = model.generate(texts=["Why are the LLM models important?"])

```

### Supported Scope

All currently supported models.

### Plan

- [x] Initial [draft PR](https://github.com/stochasticai/xTuring/pull/263) supporting [gpt2](https://github.com/yiliu30/xTuring/blob/ly/woq_support/examples/models/gpt2/gpt2_woq.py) for a quick review of the implementation details
- [x] Extend support to all models
- [x] UT and CI
- [x] Doc

@StochasticRomanAgeev @tushar2407

**Comment by StochasticRomanAgeev** [comment]
Hi @yiliu30,
Thanks for pr and interest!
First question, what is better in this approach than our already supported int8 version of models?

**Comment by yiliu30** [comment]
> Hi @yiliu30, Thanks for pr and interest! First question, what is better in this approach than our already supported int8 version of models?

Hi @StochasticRomanAgeev , thanks for your reply. there are several improvements:

- More precision options, we offer a broader range of low-precision data types, including 4 bits ([INT4, FP4, and NF4](https://github.com/intel/neural-compressor/blob/master/docs/source/quantization_weight_only.md#quantization-capability)).
- Better accuracy, we support a set of fashion algorithms, such as [RTN](https://arxiv.org/abs/2306.00978), [AWQ](https://arxiv.org/abs/2306.00978), and [TEQ](https://arxiv.org/abs/2310.10944). (some [test results](https://github.com/intel/neural-compressor/blob/master/docs/source/validated_model_list.md#pytorch-models-with-torch-201cpu-in-woq-mode))
- Faster inference, we support a set of [Highly Optimized Low Precision Kernels](https://github.com/intel/intel-extension-for-transformers/blob/main/intel_extension_for_transformers/llm/runtime/graph/core/README.md#highly-optimized-low-precision-kernels) which leverage the AMX_INT8, AVX512_VNNI, and AVX512F ISA.

**Comment by StochasticRomanAgeev** [comment]
Thanks for pr!
I have a small changes request for you -

- We need to add `intel_extension_for_transformers` as an optional dependancy.
- The way we want to use it - inside `int8` model check if we are in cpu mode - then do import and use this config. This is because we want to minimise additional steps for user to use your extension.

**Comment by StochasticRomanAgeev** [comment]
I am talking about this [if branch](https://github.com/stochasticai/xTuring/blob/55eda97e51e6b04c6796ae12104dd11cda362a47/src/xturing/engines/causal.py#L42), you need just to integrate your code there.

**Comment by yiliu30** [comment]
> Thanks for pr! I have a small changes request for you -
> 
> * We need to add `intel_extension_for_transformers` as an optional dependancy.
> * The way we want to use it - inside `int8` model check if we are in cpu mode - then do import and use this config. This is because we want to minimise additional steps for user to use your extension.

Thanks and agree with your suggestion, we will work on a new PR for it soon :)

**Comment by yiliu30** [comment]
> I am talking about this [if branch](https://github.com/stochasticai/xTuring/blob/55eda97e51e6b04c6796ae12104dd11cda362a47/src/xturing/engines/causal.py#L42), you need just to integrate your code there.

Hi @StochasticRomanAgeev, https://github.com/stochasticai/xTuring/pull/268 is the initial implementation following your suggestion, please take your time to review it.
